stew/templateutils.nim

```diff
@@ -0,0 +1,36 @@
+type CppVar[T] = distinct ptr T
+
+iterator evalTemplateParamOnceImpl[T](x: T): T =
+  yield x
+
+when defined(cpp):
+  # TODO `nim cpp` miscompiles iterators returning `var`,
+  #      so we need to emulate them in terms of pointers:
+  iterator evalTemplateParamOnceImpl[T](x: var T): CppVar[T] =
+    yield CppVar[T](addr(x))
+
+  template stripCppVar[T](p: CppVar[T]): var T =
+    ((ptr T)(p))[]
+else:
+  iterator evalTemplateParamOnceImpl[T](x: var T): var T =
+    yield x
+
+template evalTemplateParamOnce*(templateParam, newName, blk: untyped) =
```

Comment:
what's the advantage over `let x = xParam`?

Reply:
Two benefits:

* It avoids copying whenever possible.
* It works for var parameters.

Reply:
ok - I tested it with:

```
template foo(xParam: untyped) =
  evalTemplateParamOnce(xParam, x):
    echo x
    echo x

foo(@[123] & @[2134])

let v = @[432]
foo(v)

```

and in both cases, there's a temporary + copy involved  (ie also with v when in theory, it's not needed) :/

Reply:
The second example doesn't introduce a copy, so I'm not sure how you reached your conclusions. Here is the generated code:

```
	nimln_(12, "/Users/zahary/status/nimbus-eth2/vendor/nim-stew/tests/arne.nim");
	v = (tySequence__qwqHTkRvwhrRyENtudHQ7g*) newSeq((&NTI__qwqHTkRvwhrRyENtudHQ7g_), 1);
	v->data[0] = ((NI) 432);
	{
		{
			tySequence__qwqHTkRvwhrRyENtudHQ7g** paramAddrX60gensym26540001_;
			tyArray__nHXaesL0DJZHyVS07ARPRA T9_;
			tyArray__nHXaesL0DJZHyVS07ARPRA T10_;
			paramAddrX60gensym26540001_ = (tySequence__qwqHTkRvwhrRyENtudHQ7g**)0;
			nimln_(46, "/Users/zahary/status/nimbus-eth2/vendor/nim-stew/stew/templateu"
"tils.nim");
			paramAddrX60gensym26540001_ = (&v);
			nimln_(6, "/Users/zahary/status/nimbus-eth2/vendor/nim-stew/tests/arne.nim");
			nimZeroMem((void*)T9_, sizeof(tyArray__nHXaesL0DJZHyVS07ARPRA));
			T9_[0] = dollar___Gsa4WPnABj84BVdJHxUK1w((*paramAddrX60gensym26540001_));
			echoBinSafe(T9_, 1);
			nimln_(7, "/Users/zahary/status/nimbus-eth2/vendor/nim-stew/tests/arne.nim");
			nimZeroMem((void*)T10_, sizeof(tyArray__nHXaesL0DJZHyVS07ARPRA));
			T10_[0] = dollar___Gsa4WPnABj84BVdJHxUK1w((*paramAddrX60gensym26540001_));
			echoBinSafe(T10_, 1);
		}
	}
```

The key line is `paramAddrX60gensym26540001_ = (&v);` where the address of the input sequence is taken before being printed twice.

Reply:
My program was a slight modification of yours, mostly so I can more easily look at the generated code:

```nim
import
  ../stew/templateutils

template foo(xParam: untyped) =
  evalTemplateParamOnce(xParam, x):
    echo x
    echo x

proc main =
  foo(@[123] & @[2134])

  let v = @[432]
  foo(v)

main()
```

Reply:
after renaming `v` to `vvvvvv` (for clarity, as you say):

```c
    if (vvvvvv__RG7KbPR7Kab6pICXUHR9a9bg) {
      nimGCunrefNoCycle(vvvvvv__RG7KbPR7Kab6pICXUHR9a9bg);
      vvvvvv__RG7KbPR7Kab6pICXUHR9a9bg = NIM_NIL;
    }
    vvvvvv__RG7KbPR7Kab6pICXUHR9a9bg =
        (tySequence__qwqHTkRvwhrRyENtudHQ7g *)newSeqRC1(
            (&NTI__qwqHTkRvwhrRyENtudHQ7g_), 1);
    vvvvvv__RG7KbPR7Kab6pICXUHR9a9bg->data[0] = ((NI)432);
    {
      {
        tyArray__nHXaesL0DJZHyVS07ARPRA T9_;
        tyArray__nHXaesL0DJZHyVS07ARPRA T10_;
        genericSeqAssign(
            (&paramAddrX60gensym26515001___1FRaij9cAw0pBVhEf9bhR0Bw),
            vvvvvv__RG7KbPR7Kab6pICXUHR9a9bg, (&NTI__qwqHTkRvwhrRyENtudHQ7g_));
        nimZeroMem((void *)T9_, sizeof(tyArray__nHXaesL0DJZHyVS07ARPRA));
        T9_[0] = dollar___Gsa4WPnABj84BVdJHxUK1w(
            paramAddrX60gensym26515001___1FRaij9cAw0pBVhEf9bhR0Bw);
        echoBinSafe(T9_, 1);
        nimZeroMem((void *)T10_, sizeof(tyArray__nHXaesL0DJZHyVS07ARPRA));
        T10_[0] = dollar___Gsa4WPnABj84BVdJHxUK1w(
            paramAddrX60gensym26515001___1FRaij9cAw0pBVhEf9bhR0Bw);
        echoBinSafe(T10_, 1);
      }
    }
```


you can see `v` being assigned to `paramAddrX60gensym26515001___1FRaij9cAw0pBVhEf9bhR0Bw` with a `genericSeqAssign`.

I guess it's a difference of top-level vs in-proc then?

Reply:
would https://github.com/status-im/nim-stew/blob/master/stew/results.nim#L662 be a good candidate for using this function?